File: android/construction/src/main/kotlin/com/zaksecurity/construction3dbuilder/construction/WallFramer.kt
package com.zaksecurity.construction3dbuilder.construction

import com.zaksecurity.construction3dbuilder.model.CutOperation
import com.zaksecurity.construction3dbuilder.model.End
import com.zaksecurity.construction3dbuilder.model.StockFamily
import com.zaksecurity.construction3dbuilder.model.StockSpec
import com.zaksecurity.construction3dbuilder.model.Transform3D
import com.zaksecurity.construction3dbuilder.model.Vec3
import java.util.UUID
import kotlin.math.atan2
import kotlin.math.hypot


data class WallOpening(
    val type: Type,
    val offsetIn: Double,
    val widthIn: Double,
    val heightIn: Double,
    val sillHeightIn: Double = 36.0,
    val headerItemId: String = "2x8",
) {
    enum class Type { DOOR, WINDOW }
}

data class WallOptions(
    val lengthIn: Double = 96.0,
    val leftHeightIn: Double = 96.0,
    val rightHeightIn: Double = 96.0,
    val spacingOnCenterIn: Double = 16.0,
    val studItemId: String = "2x4",
    val doubleTopPlate: Boolean = true,
    val leftThreeStudCorner: Boolean = true,
    val rightThreeStudCorner: Boolean = true,
    val opening: WallOpening? = null,
)

data class WallPiece(
    val role: String,
    val spec: StockSpec,
    val transform: Transform3D,
)

data class WallPlan(
    val assemblyId: String,
    val options: WallOptions,
    val pieces: List<WallPiece>,
)

object WallFramer {
    private const val MAX_STANDARD_STOCK_IN = 192.0

    fun create(options: WallOptions = WallOptions()): WallPlan {
        val length = options.lengthIn.coerceAtLeast(12.0)
        val leftHeight = options.leftHeightIn.coerceAtLeast(12.0)
        val rightHeight = options.rightHeightIn.coerceAtLeast(12.0)
        val spacing = options.spacingOnCenterIn.coerceAtLeast(4.0)
        val stud = MaterialCatalog.lumber(options.studItemId)
        val studThickness = stud.actualThicknessIn
        val wallDepth = stud.actualWidthIn
        val topPlateCount = if (options.doubleTopPlate) 2 else 1
        val slopeAngle = Math.toDegrees(atan2(rightHeight - leftHeight, length))
        val slopedTopLength = hypot(length, rightHeight - leftHeight)
        val pieces = mutableListOf<WallPiece>()
        val assemblyId = "wall-${UUID.randomUUID()}"

        splitLength(length).forEach { segment ->
            pieces += piece(
                role = "bottom-plate",
                spec = stockSpec(options.studItemId, segment.length),
                position = Vec3(-length / 2 + segment.start + segment.length / 2, 0.0, studThickness / 2),
            )
        }

        for (layer in 0 until topPlateCount) {
            val verticalOffset = studThickness / 2 + layer * studThickness
            splitLength(slopedTopLength).forEach { segment ->
                val midpointRatio = (segment.start + segment.length / 2) / slopedTopLength
                val x = -length / 2 + length * midpointRatio
                val z = leftHeight + (rightHeight - leftHeight) * midpointRatio - verticalOffset
                pieces += piece(
                    role = if (layer == 0) "top-plate" else "top-plate-2",
                    spec = stockSpec(options.studItemId, segment.length),
                    position = Vec3(x, 0.0, z),
                    rotation = Vec3(0.0, -slopeAngle, 0.0),
                )
            }
        }

        val opening = options.opening
        var blockedStart = Double.POSITIVE_INFINITY
        var blockedEnd = Double.NEGATIVE_INFINITY
        if (opening != null) {
            val openingLeft = maxOf(studThickness * 3, opening.offsetIn)
            val openingRight = minOf(length - studThickness * 3, openingLeft + maxOf(12.0, opening.widthIn))
            blockedStart = openingLeft - studThickness * 2
            blockedEnd = openingRight + studThickness * 2
        }

        val centers = commonStudCenters(length, spacing, studThickness)
            .filter { it < blockedStart || it > blockedEnd }
            .toMutableList()

        if (options.leftThreeStudCorner) {
            centers += studThickness * 1.5
            centers += studThickness * 2.5
        }
        if (options.rightThreeStudCorner) {
            centers += length - studThickness * 1.5
            centers += length - studThickness * 2.5
        }

        fun addVerticalStud(role: String, xFromLeft: Double, bottomZ: Double = studThickness, topZ: Double? = null) {
            val wallTop = slopeHeight(leftHeight, rightHeight, length, xFromLeft)
            val actualTop = topZ ?: (wallTop - topPlateCount * studThickness)
            val cutLength = maxOf(studThickness, actualTop - bottomZ)
            val operations = if (kotlin.math.abs(slopeAngle) > 0.001 && topZ == null) {
                listOf(CutOperation.EndCut(End.B, bevelDegrees = -slopeAngle))
            } else emptyList()
            pieces += piece(
                role = role,
                spec = stockSpec(options.studItemId, cutLength, operations),
                position = Vec3(-length / 2 + xFromLeft, 0.0, bottomZ + cutLength / 2),
                rotation = Vec3(0.0, -90.0, 0.0),
            )
        }

        uniquePositions(centers).forEach { addVerticalStud("common-stud", it) }

        if (opening != null) {
            val openingLeft = maxOf(studThickness * 3, opening.offsetIn)
            val openingRight = minOf(length - studThickness * 3, openingLeft + maxOf(12.0, opening.widthIn))
            val openingWidth = openingRight - openingLeft
            val roughHeight = maxOf(24.0, opening.heightIn)
            val sillHeight = if (opening.type == WallOpening.Type.WINDOW) maxOf(studThickness * 2, opening.sillHeightIn) else studThickness
            val header = MaterialCatalog.lumber(opening.headerItemId)
            val headerBottom = minOf(
                slopeHeight(leftHeight, rightHeight, length, (openingLeft + openingRight) / 2) - topPlateCount * studThickness - header.actualWidthIn,
                if (opening.type == WallOpening.Type.DOOR) roughHeight else sillHeight + roughHeight,
            )

            val leftJack = openingLeft - studThickness / 2
            val rightJack = openingRight + studThickness / 2
            val leftKing = openingLeft - studThickness * 1.5
            val rightKing = openingRight + studThickness * 1.5

            addVerticalStud("king-stud", leftKing)
            addVerticalStud("king-stud", rightKing)
            addVerticalStud("jack-stud", leftJack, studThickness, headerBottom)
            addVerticalStud("jack-stud", rightJack, studThickness, headerBottom)

            val headerLength = openingWidth + studThickness * 2
            val headerCenterX = -length / 2 + (openingLeft + openingRight) / 2
            val headerCenterZ = headerBottom + header.actualWidthIn / 2
            val headerYOffset = maxOf(0.0, wallDepth / 2 - header.actualThicknessIn / 2)
            pieces += piece(
                "header",
                stockSpec(opening.headerItemId, headerLength),
                Vec3(headerCenterX, -headerYOffset, headerCenterZ),
                Vec3(90.0, 0.0, 0.0),
            )
            if (headerYOffset > 0.01) {
                pieces += piece(
                    "header",
                    stockSpec(opening.headerItemId, headerLength),
                    Vec3(headerCenterX, headerYOffset, headerCenterZ),
                    Vec3(90.0, 0.0, 0.0),
                )
            }

            if (opening.type == WallOpening.Type.WINDOW) {
                pieces += piece(
                    "window-sill",
                    stockSpec(options.studItemId, openingWidth),
                    Vec3(headerCenterX, 0.0, sillHeight),
                )
                commonStudCenters(length, spacing, studThickness)
                    .filter { it > openingLeft && it < openingRight }
                    .forEach { addVerticalStud("cripple-below", it, studThickness, sillHeight - studThickness / 2) }
            }

            val headerTop = headerBottom + header.actualWidthIn
            commonStudCenters(length, spacing, studThickness)
                .filter { it > openingLeft && it < openingRight }
                .forEach { addVerticalStud("cripple-above", it, headerTop) }
        }

        return WallPlan(assemblyId, options.copy(lengthIn = length, leftHeightIn = leftHeight, rightHeightIn = rightHeight, spacingOnCenterIn = spacing), pieces)
    }

    private data class Segment(val start: Double, val length: Double)

    private fun splitLength(total: Double): List<Segment> {
        val result = mutableListOf<Segment>()
        var cursor = 0.0
        while (cursor < total - 0.001) {
            val length = minOf(MAX_STANDARD_STOCK_IN, total - cursor)
            result += Segment(cursor, length)
            cursor += length
        }
        return result
    }

    private fun chooseStockLength(itemId: String, required: Double): Double {
        val candidates = MaterialCatalog.lumber(itemId).lengths.map { it.valueIn }.filter { it >= required }.sorted()
        return candidates.firstOrNull() ?: required
    }

    private fun stockSpec(itemId: String, cutLength: Double, operations: List<CutOperation> = emptyList()) = StockSpec(
        family = StockFamily.LUMBER,
        itemId = itemId,
        stockLengthIn = chooseStockLength(itemId, cutLength),
        cutLengthIn = cutLength,
        operations = operations,
    )

    private fun piece(role: String, spec: StockSpec, position: Vec3, rotation: Vec3 = Vec3()) = WallPiece(
        role = role,
        spec = spec,
        transform = Transform3D(position = position, rotationDegrees = rotation),
    )

    private fun commonStudCenters(length: Double, spacing: Double, studThickness: Double): List<Double> {
        val centers = mutableListOf(studThickness / 2)
        var x = spacing
        while (x < length - studThickness / 2) {
            centers += x
            x += spacing
        }
        centers += length - studThickness / 2
        return uniquePositions(centers)
    }

    private fun uniquePositions(values: List<Double>, tolerance: Double = 0.01): List<Double> = values.sorted().filterIndexed { index, value ->
        index == 0 || kotlin.math.abs(value - values.sorted()[index - 1]) > tolerance
    }

    private fun slopeHeight(left: Double, right: Double, length: Double, xFromLeft: Double): Double =
        if (length <= 0) left else left + (right - left) * (xFromLeft / length)
}

File: src/editor/wall-framer.js
import { getLumber } from '../model/material-catalog.js';

function chooseStockLength(itemId, required) {
  const item = getLumber(itemId);
  const candidates = item.lengths
    .map(option => Number(option.value))
    .filter(value => value >= required)
    .sort((a, b) => a - b);
  return candidates[0] || Math.max(required, ...item.lengths.map(option => Number(option.value)));
}

function stockSpec(itemId, cutLength, operations = []) {
  return {
    kind: 'construction',
    family: 'lumber',
    itemId,
    stockLength: chooseStockLength(itemId, cutLength),
    cutLength,
    operations,
  };
}

function part(role, spec, position, rotation = [0, 0, 0]) {
  return { role, spec, position, rotation };
}

function uniquePositions(values, tolerance = 0.01) {
  return values
    .sort((a, b) => a - b)
    .filter((value, index, array) => index === 0 || Math.abs(value - array[index - 1]) > tolerance);
}

function commonStudCenters(length, spacing, studThickness) {
  const centers = [studThickness / 2];
  for (let x = spacing; x < length - studThickness / 2; x += spacing) centers.push(x);
  centers.push(length - studThickness / 2);
  return uniquePositions(centers);
}

function slopeHeight(leftHeight, rightHeight, length, xFromLeft) {
  if (length <= 0) return leftHeight;
  return leftHeight + (rightHeight - leftHeight) * (xFromLeft / length);
}

function splitPlate(length, maxLength = 192) {
  const segments = [];
  let remaining = length;
  let cursor = 0;
  while (remaining > 0.001) {
    const segmentLength = Math.min(maxLength, remaining);
    segments.push({ start: cursor, length: segmentLength });
    cursor += segmentLength;
    remaining -= segmentLength;
  }
  return segments;
}

export function createWallPlan(options = {}) {
  const length = Math.max(12, Number(options.length) || 96);
  const leftHeight = Math.max(12, Number(options.leftHeight) || 96);
  const rightHeight = Math.max(12, Number(options.rightHeight) || leftHeight);
  const spacing = Math.max(4, Number(options.spacing) || 16);
  const studItemId = options.studItemId || '2x4';
  const stud = getLumber(studItemId);
  const studThickness = stud.actual[0];
  const wallDepth = stud.actual[1];
  const topPlateCount = options.doubleTop === false ? 1 : 2;
  const slopeAngle = Math.atan2(rightHeight - leftHeight, length) * 180 / Math.PI;
  const slopedTopLength = Math.hypot(length, rightHeight - leftHeight);
  const pieces = [];
  const assemblyId = options.assemblyId || `wall-${Date.now()}`;

  // Bottom plate(s), split at the longest catalog stock length if necessary.
  for (const segment of splitPlate(length)) {
    pieces.push(part(
      'bottom-plate',
      stockSpec(studItemId, segment.length),
      [-length / 2 + segment.start + segment.length / 2, 0, studThickness / 2],
    ));
  }

  // Sloped or level top plates. They stay individual boards for takeoff and editing.
  for (let layer = 0; layer < topPlateCount; layer += 1) {
    const normalOffset = studThickness / 2 + layer * studThickness;
    pieces.push(part(
      layer === 0 ? 'top-plate' : 'top-plate-2',
      stockSpec(studItemId, slopedTopLength),
      [0, 0, (leftHeight + rightHeight) / 2 - normalOffset],
      [0, -slopeAngle, 0],
    ));
  }

  const opening = options.opening && options.opening.type !== 'none' ? options.opening : null;
  let blockedStart = Infinity;
  let blockedEnd = -Infinity;
  if (opening) {
    const openingLeft = Math.max(studThickness * 3, Number(opening.offset) || 24);
    const openingWidth = Math.max(12, Number(opening.width) || 36);
    const openingRight = Math.min(length - studThickness * 3, openingLeft + openingWidth);
    blockedStart = openingLeft - studThickness * 2;
    blockedEnd = openingRight + studThickness * 2;
  }

  const centers = commonStudCenters(length, spacing, studThickness)
    .filter(center => center < blockedStart || center > blockedEnd);

  if (options.leftCorner !== false) {
    centers.push(studThickness * 1.5, studThickness * 2.5);
  }
  if (options.rightCorner !== false) {
    centers.push(length - studThickness * 1.5, length - studThickness * 2.5);
  }

  function addVerticalStud(role, xFromLeft, bottomZ = studThickness, topZ = null) {
    const wallTop = slopeHeight(leftHeight, rightHeight, length, xFromLeft);
    const topPlateAllowance = topPlateCount * studThickness;
    const actualTop = topZ == null ? wallTop - topPlateAllowance : topZ;
    const cutLength = Math.max(studThickness, actualTop - bottomZ);
    const operations = Math.abs(slopeAngle) > 0.001 && topZ == null
      ? [{ type: 'endCut', end: 'B', miter: 0, bevel: -slopeAngle }]
      : [];
    pieces.push(part(
      role,
      stockSpec(studItemId, cutLength, operations),
      [-length / 2 + xFromLeft, 0, bottomZ + cutLength / 2],
      [0, -90, 0],
    ));
  }

  for (const center of uniquePositions(centers)) addVerticalStud('common-stud', center);

  if (opening) {
    const openingLeft = Math.max(studThickness * 3, Number(opening.offset) || 24);
    const requestedWidth = Math.max(12, Number(opening.width) || 36);
    const openingRight = Math.min(length - studThickness * 3, openingLeft + requestedWidth);
    const openingWidth = openingRight - openingLeft;
    const roughHeight = Math.max(24, Number(opening.height) || 82.5);
    const sillHeight = opening.type === 'window' ? Math.max(studThickness * 2, Number(opening.sillHeight) || 36) : studThickness;
    const headerItemId = opening.headerItemId || '2x6';
    const header = getLumber(headerItemId);
    const headerThickness = header.actual[0];
    const headerDepth = header.actual[1];
    const headerBottom = Math.min(
      slopeHeight(leftHeight, rightHeight, length, (openingLeft + openingRight) / 2) - topPlateCount * studThickness - headerDepth,
      opening.type === 'door' ? roughHeight : sillHeight + roughHeight,
    );

    const leftJack = openingLeft - studThickness / 2;
    const rightJack = openingRight + studThickness / 2;
    const leftKing = openingLeft - studThickness * 1.5;
    const rightKing = openingRight + studThickness * 1.5;

    addVerticalStud('king-stud', leftKing);
    addVerticalStud('king-stud', rightKing);
    addVerticalStud('jack-stud', leftJack, studThickness, headerBottom);
    addVerticalStud('jack-stud', rightJack, studThickness, headerBottom);

    const headerLength = openingWidth + studThickness * 2;
    const headerCenterX = -length / 2 + (openingLeft + openingRight) / 2;
    const headerCenterZ = headerBottom + headerDepth / 2;
    const headerYOffset = Math.max(0, wallDepth / 2 - headerThickness / 2);
    pieces.push(part(
      'header',
      stockSpec(headerItemId, headerLength),
      [headerCenterX, -headerYOffset, headerCenterZ],
      [90, 0, 0],
    ));
    if (headerYOffset > 0.01) {
      pieces.push(part(
        'header',
        stockSpec(headerItemId, headerLength),
        [headerCenterX, headerYOffset, headerCenterZ],
        [90, 0, 0],
      ));
    }

    if (opening.type === 'window') {
      pieces.push(part(
        'window-sill',
        stockSpec(studItemId, openingWidth),
        [headerCenterX, 0, sillHeight],
      ));

      for (const center of commonStudCenters(length, spacing, studThickness)) {
        if (center > openingLeft && center < openingRight) {
          addVerticalStud('cripple-below', center, studThickness, sillHeight - studThickness / 2);
        }
      }
    }

    const headerTop = headerBottom + headerDepth;
    for (const center of commonStudCenters(length, spacing, studThickness)) {
      if (center > openingLeft && center < openingRight) {
        addVerticalStud('cripple-above', center, headerTop);
      }
    }
  }

  return {
    assemblyId,
    type: 'wall',
    length,
    leftHeight,
    rightHeight,
    spacing,
    studItemId,
    pieces,
  };
}
